### ISSUE TITLE
How to make clusters more compact?

### ISSUE THREAD

**Issue by claczny** [KlugerLab/FIt-SNE#75]
Dear all,

first of all, thanks for the great tool! I am currently developing upon it and it has been really nice to use so far.

When comparing it to "classical" t-SNE visualizations, I found the classic visualization (i.e., the 2D embedded coordinates) to be somewhat more compact. Put differently, the clusters were more strongly separated.
This, in turn, is relevant as I would like to cluster the 2D embedded data.

Could you please let me know if and, if so, how this could be approached using FIt-SNE?
I checked the documentation and it seems that the late exageration parameter might be useful here.
However, I have no intuition what would be a good cooefficient there to start with and at which iteration to start.

Your thoughts and input are greatly appreciated!

Best,

Cedric

P.S. I am currently using v.1.1.0 and the python wrapper.

**Comment by dkobak** [comment]
What exactly do you mean by "classical" t-SNE here? Can you show a difference between "classical" t-SNE and FIt-SNE result? There should not be much difference.

Apart from that -- what is the size of your data set? This can help making further suggestions.

**Comment by claczny** [comment]
Thank you for the quick reply.

> What exactly do you mean by "classical" t-SNE here? Can you show a difference between "classical" t-SNE and FIt-SNE result? There should not be much difference.

I see. It could be that it is then a downstream "effect" of how we implemented this in [VizBin](https://github.com/claczny/VizBin). I will check how I can best provide the different outcomes here.

> Apart from that -- what is the size of your data set? This can help making further suggestions.

The size of the data set is variable. It can range from a few hundred to several tens of thousands (maybe hundred thousands) of points.
The dimensionality is typically between 100 - 1000 dimensions.

**Comment by dkobak** [comment]
For small data sets (let's say n<10,000 or so) you can try larger perplexities. For large data sets you can try exaggeration. I recommend turning exaggeration on directly after the early exaggeration phase, e.g. `late_exag_coeff=2, start_late_exag_iter=250`. Reasonable values to try out are 2 or 4. You might be interested in reading our preprint https://www.biorxiv.org/content/10.1101/453449v2 for further discussion. Cheers.

**Comment by claczny** [comment]
I see, thank you!

If this worked, I now attached two files.
[vizbin.txt](https://github.com/KlugerLab/FIt-SNE/files/3226439/vizbin.txt) contains the "classic" coordinates (29212x2, i.e., 29212 datapoints) which, IMO, have more "compact" clusters.
[fitsne.txt](https://github.com/KlugerLab/FIt-SNE/files/3226448/fitsne.txt) contains what I am currently working on and while the clusters are clearly apparent, it would be great if they could be less "blurry".

I will try the late exaggeration.

**Comment by dkobak** [comment]
I suggest you post figures instead.

Also, I'm still not sure what you mean by "classic". Barnes-Hut? Which implementation?

**Comment by claczny** [comment]
> Also, I'm still not sure what you mean by "classic". Barnes-Hut? Which implementation?

I meant the BH-SNE implementation from Laurens.

I tried a `late_exag_coef=1.3` and it seems to make a large difference and absolutely in the right direction!

**Comment by dkobak** [comment]
Note also that whatever approximation method you are using (BH or FI), it makes sense to scale the learning rate with sample size, e.g. set learning rate to n/12. See the preprint linked above and references therein. If your n is above ~100,000, you might see a substantial effect (compared to the default learning rate 200).

**Comment by claczny** [comment]
@dkobak I have not found the time yet to read your preprint but I remembered that you mention "PCA" in the abstract.
So I checked what difference running a PCA from high-D to 50D *prior* to FIt-SNE would make for the data that I am working on and it results in a substantial improvement!
Combined with the `late_exag_coeff` the clusters have a very nice 2D-Gaussian shape and are nicely separated which is convenient for automated clustering. 
Put differently, the data looks now much more like I was used to when using Laurens' BHt-SNE implementation but at ~1/7 the compute time!

**Comment by dkobak** [comment]
This is a good point. I just checked and bhtsne wrappers (at least the MATLAB one) do PCA by default, whereas fitsne wrappers currently don't.

This makes me wonder if we should add PCA to the wrappers to preserve the default behaviour. George?

**Comment by linqiaozhi** [comment]
Hi @dkobak and @claczny,

As Dmitry pointed out, there should generally not be an appreciable difference between the embeddings produced by FIt-SNE and BH t-SNE. However, Dmitry is absolutely correct that the BH t-SNE has a built in PCA step by default.

> This makes me wonder if we should add PCA to the wrappers to preserve the default behaviour. George?

That's a good question. Although I recognize the appeal of doing the same with FIt-SNE (primarily that the interface be consistent with BH t-SNE), I am pretty hesitant to change FIt-SNE to do PCA by default. I think PCA should be deliberately done prior to t-SNE (when it is done at all). This allows the user to be clear as to how many PCs they are retaining (why should default be 50, like in BH t-SNE?) and which algorithm they use (I almost always use randomized SVD, but there are circumstances where Lanczos methods, or even exact PCA might be more appropriate). For PCA of big datasets that FIt-SNE is expected to work on, you really need randomized or Lanczos, which adds an additional dependency. 

Also, I think that most people (for example, me when I first started using t-SNE) would not expect that by default it also does PCA to 50 components. For example, I remember passing the PCs (more than 50) into the t-SNE wrapper and I was surprised to find that it was doing a super slow PCA (to a smaller number of PCs) before it even started doing t-SNE. PCA is often a good idea before t-SNE, but it is certainly not the first step of t-SNE by any definition of the latter. Finally, such a change would break almost all code that depends on FIt-SNE (e.g. literally all of my code). 

Do you have a different view, @dkobak?

**Comment by dkobak** [comment]
> Finally, such a change would break almost all code that depends on FIt-SNE (e.g. literally all of my code).

Fair enough. I guess it's too late to change the default behaviour.

I'd suggest to add a warning to the wrappers: if the input dimensionality is above 50, then it prints NOTE THAT WE ARE NOT DOING PCA (well, something to that effect).

Also, we could add *optional* PCA, turned off by default.

> there are circumstances where Lanczos methods, or even exact PCA 

Isn't Lanczos exact?

**Comment by linqiaozhi** [comment]
> I'd suggest to add a warning to the wrappers: if the input dimensionality is above 50, then it prints NOTE THAT WE ARE NOT DOING PCA (well, something to that effect).

Agreed. Although perhaps the threshold should be higher (e.g. 200), as if you only have 50 dimensions then it might be reasonable to not do PCA first? 

> Also, we could add _optional_ PCA, turned off by default.

I would definitely be open to that.

> 
> > there are circumstances where Lanczos methods, or even exact PCA
> 
> Isn't Lanczos exact?

In exact arithmetic, Lanczos methods converge to the correct eigenvectors. But even modern implementations can be numerically unstable. When computing the top eigenvectors of a low-rank matrix, randomized methods are faster and more stable. For example, you can check out our [paper](https://dl.acm.org/citation.cfm?doid=2988516.3004053) from a few years ago comparing an implementation of randomized SVD to ARPACK.

**Comment by dkobak** [comment]
> Although perhaps the threshold should be higher (e.g. 200), as if you only have 50 dimensions then it might be reasonable to not do PCA first?

I was just thinking about a legacy warning, to warn about a difference compared to BH implementations.

I'll reply about Lanczos later; will check the paper.

**Comment by dkobak** [comment]
@linqiaozhi 

> When computing the top eigenvectors of a low-rank matrix, randomized methods are faster and more stable. For example, you can check out our paper from a few years ago comparing an implementation of randomized SVD to ARPACK.

This is interesting. Until now, I was under impression that `svds` returns the results equivalent to `svd` up to numerical error (or very close to it), whereas randomized svd, at least as it is implemented in `scikit-learn` with default parameters, is differing from `svd` much more strongly. I had a related discussion some time ago here: https://github.com/theislab/scanpy/issues/325:

> Regarding the result: the high PCs [in randomized PCA] can change drastically depending on the platform and the random seed. I've seen clustering results changing completely after I became aware of it.

and I've been preferring `svd` since then...
